# gin

gin (GWAS Incorporating Networks) is a software package aimed at improving biomarker discovery on genotyping data using a priori information, namely networks. It is the successor of [SConES](http://bioinformatics.oxfordjournals.org/content/29/13/i171.short), the network guided multi-locus mapping method.

*gin* is based on [easyGWAS](http://easygwas.ethz.ch), a framework for computing genome-wide association studies and meta-analysis developed by [dominikgrimm](https://github.com/dominikgrimm). This is a C/C++ Framework with Python interfaces. The code includes several standard methods for performing GWAS, such as linear regression, logistic regression and popular linear mixed models ([EMMAX] (http://www.nature.com/ng/journal/v42/n4/abs/ng.548.html), [FaSTLMM](http://www.nature.com/nmeth/journal/v8/n10/abs/nmeth.1681.html)) to also account for population stratification.

## Tutorials

Tutorials can be found in the Wiki section of dominikgrimmÂ´s repository: http://github.com/dominikgrimm/easyGWASCore/wiki
